src/polyswarmd/rpc.py

```diff
@@ -0,0 +1,254 @@
+from concurrent.futures import ThreadPoolExecutor
+import gevent
+import json
+
+from gevent.lock import BoundedSemaphore
+from requests.exceptions import ConnectionError
+from polyswarmartifact.schema import Bounty as BountyMetadata
+from requests_futures.sessions import FuturesSession
+
+from polyswarmd.utils import *
+
+
+class EthereumRpc:
+    """
+    This class periodically polls several geth filters, and multicasts the results across any open WebSockets
+    """
+    def __init__(self, chain):
+        self.chain = chain
+        self.session = FuturesSession(executor=ThreadPoolExecutor(32),
+                                      adapter_kwargs={'max_retries': 3})
+        self.assertion_filter = None
+        self.block_filter = None
+        self.bounty_filter = None
+        self.deprecated_filter = None
+        self.fee_filter = None
+        self.init_filter = None
+        self.quorum_filiter = None
+        self.reveal_filter = None
+        self.settled_filter = None
+        self.vote_filter = None
+        self.window_filter = None
+        self.websockets_lock = BoundedSemaphore(1)
+        self.websockets = None
+
+    def broadcast(self, message):
+        """
+        Send a message to all connected WebSockets
+        :param message: dict to be converted to json and sent
+        """
+        logger.debug('Sending: %s', message)
+        with self.websockets_lock:
+            for ws in self.websockets:
+                logger.debug('Sending WebSocket %s %s', ws, message)
+                ws.send(json.dumps(message))
+
+    def flush_filters(self):
+        """
+        Clear filters of existing entires
+        """
+        self.block_filter.get_new_entries()
+        self.fee_filter.get_new_entries()
+        self.window_filter.get_new_entries()
+        self.bounty_filter.get_new_entries()
+        self.assertion_filter.get_new_entries()
+        self.vote_filter.get_new_entries()
+        self.quorum_filiter.get_new_entries()
+        self.settled_filter.get_new_entries()
+        self.reveal_filter.get_new_entries()
+        self.deprecated_filter.get_new_entries()
+        if self.init_filter:
+            self.init_filter.get_new_entries()
+
+    # noinspection PyBroadException
+    def poll(self, ipfs_uri):
+        """
+        Continually poll all Ethereum filters as long as there are WebSockets listening
+        :param ipfs_uri: IPFS root uri
+        """
+        self.setup_filters()
+        from polyswarmd.bounties import substitute_ipfs_metadata
+        while True:
+            gevent.sleep(1)
```

Comment:
This sleep really bothers me. Once we have a ton of stuff to dump to the socket, doesn't this artificially slow down the broadcast loop? Could we instead keep track of the time when the broadcast started/ended and sleep that amount of time?

Something like:

```
last = time.time()
while True:
   current_time = time.time()
    sleep = 1-current_time-last
    gevent.sleep(sleep if sleep > 0 else 0)
    last = current_time
```

That way we always yield at least once per iteration so other tasks don't get starved under load, but only sleep if we actually don't have much to do.

Reply:
Since we monkey patched the socket with gevent, we could probably completely remove the sleep & it will operating asynchronously on network io.

Reply:
It would be great to try that.